### PR TITLE
attack viewchange [Idea| WIP]

### PR DESCRIPTION
**Pull request by alajko**
## Issue

To enable functionality to attack the consensus mechanism in general and view-change in particular.
Currently implemented `delayResponse` attack.

## TODO

- [ ] Merge new deploy-malicious.sh script into deploy.sh
- [ ] Cleanup old attack mode code.
- [ ] Implement other attack models

### CODE REVIEW

**Comment by LeoHChen** [comment]
This PR had caused panic crash. The log is available at, s3://harmony-benchmark/logs/2019/05/18/235808/

**Comment by alajko** [comment]
accidentally deleted @LeoHChen comment It said: "This PR had caused panic crash. The log is available at, s3://harmony-benchmark/logs/2019/05/18/235808/"
This was because code change affected all validators and all of them crashed, showing many panic messages. This has been now fixed.

**Comment by LeoHChen** [comment]
Is this a complete PR?  Doesn't seem like the code is complete.

**Comment by alajko** [comment]
Its not complete, it says WIP in the title @LeoHChen

**Comment by alajko** [comment]
The idea is to get initial comments on direction of the PR. @LeoHChen

File: internal/attack/attack.go [hunk 1]
```diff
@@ -63,6 +63,13 @@ func (attack *Model) SetAttackEnabled(AttackEnabled bool) {
 	}
 }
 
+// SetDelayResponseAttack sets attack to DelayResponse enabled.
+func (attack *Model) SetDelayResponseAttack() {
+	attack.AttackEnabled = true
+	attack.attackType = Type(1)
```
Comment (mikedoan) [inline internal/attack/attack.go]: why not using DelayResponse for Type(1)

File: cmd/harmony/main.go [hunk 2]
```diff
@@ -217,6 +219,11 @@ func setUpConsensusAndNode(nodeConfig *nodeconfig.ConfigType) (*consensus.Consen
 	// TODO: consensus object shouldn't start here
 	// TODO(minhdoan): During refactoring, found out that the peers list is actually empty. Need to clean up the logic of consensus later.
 	currentConsensus, err := consensus.New(nodeConfig.Host, nodeConfig.ShardID, nodeConfig.Leader, nodeConfig.ConsensusPriKey)
+	attackInstance := attack.GetInstance()
+	if *delayAttack {
+		attackInstance.SetDelayResponseAttack()
+	}
+	currentConsensus.DelayAttack = attackInstance
```
Comment (mikedoan) [inline cmd/harmony/main.go]: i don't think you need to introduce a new object in consensus - attack.GetInstance() is a singleton which can be called anywhere.

File: cmd/harmony/main.go [hunk 3]
```diff
@@ -217,6 +219,11 @@ func setUpConsensusAndNode(nodeConfig *nodeconfig.ConfigType) (*consensus.Consen
 	// TODO: consensus object shouldn't start here
 	// TODO(minhdoan): During refactoring, found out that the peers list is actually empty. Need to clean up the logic of consensus later.
 	currentConsensus, err := consensus.New(nodeConfig.Host, nodeConfig.ShardID, nodeConfig.Leader, nodeConfig.ConsensusPriKey)
+	attackInstance := attack.GetInstance()
+	if *delayAttack {
+		attackInstance.SetDelayResponseAttack()
```
Comment (mikedoan) [inline cmd/harmony/main.go]: no need to create an instance - just use attack.GetInstance().SetDelayResponseAttack()

File: consensus/consensus.go [hunk 4]
```diff
@@ -148,6 +149,7 @@ type Consensus struct {
 
 	// Staking information finder
 	stakeInfoFinder StakeInfoFinder
+	DelayAttack     *attack.Model
```
Comment (mikedoan) [inline consensus/consensus.go]: as having singleton, we don't have to introduce a pointer here.

File: consensus/consensus_validator.go [hunk 5]
```diff
@@ -129,7 +129,10 @@ func (consensus *Consensus) processPreparedMessage(message *msg_pb.Message) {
 	//#### END Read payload data
 
 	// Update readyByConsensus for attack.
-	attack.GetInstance().UpdateConsensusReady(consensusID)
+	if consensus.DelayAttack.AttackEnabled {
+		consensus.DelayAttack.UpdateConsensusReady(consensusID)
```
Comment (mikedoan) [inline consensus/consensus_validator.go]: Call singleton

File: consensus/consensus_validator_test.go [hunk 6]
```diff
@@ -102,6 +103,7 @@ func TestProcessMessageValidatorAnnounce(test *testing.T) {
 	}
 
 	consensusValidator1, err := New(m, 0, leader, bls_cosi.RandPrivateKey())
+	consensusValidator1.DelayAttack = attack.GetInstance()
```
Comment (mikedoan) [inline consensus/consensus_validator_test.go]: let's call singleton

**Pull request closed without merging**